### DISCUSSION TITLE
C++ implementation

### DISCUSSION THREAD

**Discussion by brj0** [lmcinnes/pynndescent#223] [Show and tell]
Hi everyone,

I'd like to introduce [nndescent](https://github.com/brj0/nndescent), a C++ implementation of the pynndescent project with Python bindings.

Here are some key highlights of nndescent:

**The + Side:**

- The usage is almost identical to pynndescent.
- Faster Import: Import is very fast compared to pynndescent, which uses numba real-time compilation.
- Faster Index Building: The index build process is quicker compared to the Python implementation.
- Sparse Version Performance: On my platform, the sparse version of nndescent showcases notable performance improvements.

**The - Side:**

- Benchmark Query Time: While nndescent excels in build time, at the moment the pure query time (after index building and query function preparation) on ANN benchmarks is still slower than pynndescent.
- Until now, nndescent has been developed and tested only on Linux.

I want to emphasize that I'm not a professional programmer; rather this project was undertaken during my free time as a learning experience in C++ with Python bindings. So I assume that there is still much room for improvement regarding my C++ code.

The project is **licensed under BSD2**. It took a considerable amount of effort to complete, and I hope that someone may find it useful. I welcome any contributions, pull requests, forks or integration into your own projects.

You can find the nndescent repository at this link: [https://github.com/brj0/nndescent](https://github.com/brj0/nndescent)

Feel free to reach out with any feedback, questions, or suggestions.

Best regards
Jon

**Comment by jlmelville** [comment]
As someone who also wrote a C++ implementation of NND but kept getting distracted by other things before getting all the features in I wanted, I salute you for getting it over the finish line. Looking forward to checking this out. Congrats.

**Comment by Ghazi-Kazmi** [comment]
> 

😯
